caikit/runtime/wrappers/pydantic_wrapper.py

```diff
@@ -0,0 +1,162 @@
+    BoolSequence,
+    FloatSequence,
+    IntSequence,
+    StrSequence,
+)
+
+# PYDANTIC_TO_DM_MAPPING is essentially a 2-way map of DMs <-> Pydantic models, you give it a
+# pydantic model, it gives you back a DM class, you give it a
+# DM class, you get back a pydantic model.
+PYDANTIC_TO_DM_MAPPING = {
+    # Map primitive sequences to lists
+    StrSequence: List[str],
+    IntSequence: List[int],
+    FloatSequence: List[float],
+    BoolSequence: List[bool],
+}
+
+
+# Base class for pydantic models
+# We want to set the config to forbid extra attributes
+# while instantiating any pydantic models
+# This is done to make sure any oneofs can be
+# correctly infered by pydantic
+class ParentPydanticBaseModel(pydantic.BaseModel):
+    model_config = pydantic.ConfigDict(extra="forbid")
+
+
+def pydantic_to_dataobject(pydantic_model: pydantic.BaseModel) -> DataBase:
+    """Convert pydantic objects to our DM objects"""
+    dm_class_to_build = PYDANTIC_TO_DM_MAPPING.get(type(pydantic_model))
+    dm_kwargs = {}
+
+    for field_name, field_value in pydantic_model:
+        # field could be a DM:
+        # pylint: disable=unidiomatic-typecheck
+        if type(field_value) in PYDANTIC_TO_DM_MAPPING:
+            dm_kwargs[field_name] = pydantic_to_dataobject(field_value)
+        elif isinstance(field_value, list):
+            if all(type(val) in PYDANTIC_TO_DM_MAPPING for val in field_value):
+                dm_kwargs[field_name] = [
+                    pydantic_to_dataobject(val) for val in field_value
+                ]
+            else:
+                dm_kwargs[field_name] = field_value
+        else:
+            dm_kwargs[field_name] = field_value
+
+    return dm_class_to_build(**dm_kwargs)
+
+
+def dataobject_to_pydantic(dm_class: Type[DataBase]) -> Type[pydantic.BaseModel]:
+    """Make a pydantic model based on the given proto message by using the data
+    model class annotations to mirror as a pydantic model
+    """
+    # define a local namespace for type hints to get type information from.
+    # This is needed for pydantic to have a handle on JsonDict and JsonDictValue while
+    # creating its base model
+    localns = {"JsonDict": dict, "JsonDictValue": dict}
+
+    if dm_class in PYDANTIC_TO_DM_MAPPING:
+        return PYDANTIC_TO_DM_MAPPING[dm_class]
+
+    annotations = {
+        field_name: _get_pydantic_type(field_type)
+        for field_name, field_type in get_type_hints(dm_class, localns=localns).items()
+    }
+    pydantic_model = type(ParentPydanticBaseModel)(
+        dm_class.get_proto_class().DESCRIPTOR.full_name,
+        (ParentPydanticBaseModel,),
+        {
+            "__annotations__": annotations,
+            **{
+                name: None
+                for name, _ in get_type_hints(
+                    dm_class,
+                    localns=localns,
+                ).items()
+            },
+        },
+    )
+    PYDANTIC_TO_DM_MAPPING[dm_class] = pydantic_model
```

Comment:
NIT: Now that this is in a free function and _might_ conceivably be used somewhere besides the `http_server`, this kind of global state mutation worries me a bit. If we keep this in a generic module like `runtime.wrappers`, I think we should at least put a `threading.Lock` around access to the dict. If we move this into a `runtime.http_server` module, then I think it's fine to leave unlocked since we _know_ it won't be used from a multithreaded context.

Reply:
Sounds good!